Add support for replace/original to C# parser; add SourceGenerator tests; incorporate PR feedback

cston opened the pull request:
Add support for replace/original to C# parser; add tests; and incorporate PR feedback from 56136680f52ea022a4d93dc7bb5fe69ea1c5dbdf.

@dotnet/roslyn-compiler, @dotnet/roslyn-ide, @CyrusNajmabadi, @mattwar please review

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -1462,6 +1418,63 @@ private void ParseModifiers(SyntaxListBuilder tokens)
+        /// <summary>
+        /// Return true if the current contextual keyword should be treated as a modifier.
+        /// </summary>
+        private bool IsContextualKeywordActualModifier()
```
I feel like the name would be better if it matched your comment.  i.e. "CurrentContextualKeywordShoudlBeTreatedAsAModifier()"

cston replied:
Renamed `ShouldCurrentContextualKeywordBeTreatedAsModifier`.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs:
```diff
@@ -1168,7 +1168,7 @@ internal override bool CanAddProjectReference(ProjectId referencingProject, Proj
         /// forwards the calls down to the underlying Workspace.
         /// </summary>
-        private sealed class VisualStudioWorkspaceHost : IVisualStudioWorkspaceHost, IVisualStudioWorkingFolder
+        protected class VisualStudioWorkspaceHost : IVisualStudioWorkspaceHost, IVisualStudioWorkingFolder
```
why?

cston replied:
This is reverting a possibly incorrect change that you noticed in the previous PR.

AlekseyTs commented on src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:
```diff
@@ -376,7 +372,30 @@ private int RunCore(TextWriter consoleOutput, ErrorLogger errorLogger, Cancellat
+                    {
+                        compilation = compilation.AddSyntaxTrees(trees);
+                        // Are there parse warnings? If so, we're probably reporting warnings
+                        // from the previous GetParseDiagnostics() call again here.
```
Does this need a follow-up? Do we have an issue tracking this work? I suppose it could be a TODO comment without an issue instead.

cston replied:
Changed comment to a TODO.

gafter commented:
In LanguageParser.cs, reset points do not appear to save and restore `_syntaxFactoryContext.IsInReplace`

AlekseyTs commented:
Also should investigate if adding /removing the ```replace``` modifier should be specially tracked by an incremental parser.

gafter commented on src/Compilers/CSharp/Portable/Syntax/Syntax.xml:
```diff
@@ -643,6 +643,21 @@
       <summary>Creates an BaseExpressionSyntax node.</summary>
     </FactoryComment>
   </Node>
+  <Node Name="OriginalExpressionSyntax" Base="InstanceExpressionSyntax">
```
An alternate implementation would be to just have `original` be an ordinary identifier token, like `value` in a property setter, and let semantic analysis take care of it (i.e. bind it to a method).

cston replied:
@gafter I'll keep that approach in mind. For now, this follows the approach used for `AwaitExpressionSyntax`.

cston commented:
@gafter, @alekseyts Thanks for identifying the incremental parsing issues. I've updated `GetResetPoint()`  and `Reset()` and added some incremental parsing tests. For now, those tests are failing.

cston commented:
@gafter, @alekseyts, @vsadov Updated PR to include binding and metadata writing of replaced members. Replaced members are now represented in the containing type similar to partial methods - the replacing member is included in `GetMembers()` and the replaced member is a property on the replacing member. Note that a number of the binding tests for `replace ... original` fail currently.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs:
```diff
@@ -513,6 +513,20 @@ private Symbol GetMemberSymbol(string memberName, TextSpan memberSpan, NamedType
                         }
+
+                        var method = (MethodSymbol)sym;
+                        while (true)
```
Consider adding a comment explaining why do we need the loop.

cston replied:
Added.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -1637,6 +1639,70 @@ private BoundExpression BindCastCore(ExpressionSyntax node, BoundExpression oper
+            {
+                case SymbolKind.Method:
+                    {
+                        var replacedMethod = (MethodSymbol)replaced;
```
In case of generic method, I think we should construct it with the current method type arguments/parameters.

cston replied:
Good catch. Fixed.

gafter replied:
@cston Are you updating this PR with the fixes?

cston replied:
Yes. For this particular issue, the fix is to pass `replacedByMethod.TypeArguments` at line 1665.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -1637,6 +1639,70 @@ private BoundExpression BindCastCore(ExpressionSyntax node, BoundExpression oper
+                            lookupError: null,
+                            flags: BoundMethodGroupFlags.None,
+                            receiverOpt: receiver,
+                            resultKind: LookupResultKind.Viable);
```
Do we want subsequent access to the replaced method to be virtual or non-virtual?

mattwar replied:
accessing the member via replace should not be a virtual access since it should always access exactly the replaced member, not an overload.

AlekseyTs replied:
Then we should probably make sure SuppressVirtualCalls is true for the receiver.

cston replied:
The replaced method is no longer virtual. Regardless, the call doesn't need to be virtual so emit now generates `Call` rather than `CallVirt` for replaced methods.

AlekseyTs replied:
IOperation pulls this information from BoundTree, so, if the symbol is virtual, it will likely to treat this call as virtual.

cston replied:
Updated `IInvocationExpression` implementation to return false for `IsVirtual` for replaced method calls.

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -4227,9 +4227,6 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   </data>
-  <data name="MustCallSetMethodTestData" xml:space="preserve">
-    <value>Must call SetMethodTestData(ConcurrentDictionary) before calling SetMethodTestData(MethodSymbol, ILBuilder)</value>
-  </data>
```
Is this change feature related?

cston replied:
The changes around `CompilationTestData` are not specific to this feature but are needed here to allow multiple methods with the same signature.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceEventSymbol.cs:
```diff
@@ -326,6 +326,31 @@ public sealed override bool IsVirtual
+        }
+
+        private SourceEventSymbol _replacedBy;
+        private SourceEventSymbol _replaced;
```
Please declare all fields at the top of the type declaration, before all the constructors.

cston replied:
Moved.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs:
```diff
@@ -669,6 +675,36 @@ internal sealed override OverriddenOrHiddenMembersResult OverriddenOrHiddenMembe
+        }
+
+        private SourceMethodSymbol _replacedBy;
+        private SourceMethodSymbol _replaced;
```
Please move to the top of type declaration.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs:
```diff
@@ -574,6 +574,31 @@ internal bool IsNew
+        }
+
+        private SourcePropertySymbol _replacedBy;
+        private SourcePropertySymbol _replaced;
```
Please move to the top of type declaration.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs:
```diff
@@ -574,6 +574,31 @@ internal bool IsNew
+        internal sealed override bool IsReplace
+        {
+            get { return (_modifiers & DeclarationModifiers.Replace) != 0; }
+        }
```
Should we override MetadataName?

AlekseyTs replied:
And for events?

cston replied:
Replaced properties and events are not emitted to metadata, only the accessors are.

cston commented:
All feedback has been addressed in most recent commits.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/Expression.cs:
```diff
@@ -42,7 +39,10 @@ internal partial class BoundCall : IInvocationExpression
+        bool IInvocationExpression.IsVirtual =>
+            (this.Method.IsVirtual || this.Method.IsAbstract || this.Method.IsOverride) &&
+            ((object)this.Method.ReplacedBy == null) &&
+            !this.ReceiverOpt.SuppressVirtualCalls;
```
I think there are two more similar places in this file.

cston replied:
Thanks. Fixed.

gafter commented on src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:
```diff
@@ -437,6 +438,12 @@ private void CompileNamedType(NamedTypeSymbol containingType)
+                            // Replaced methods are not included in GetMembers().
+                            foreach (var replaced in method.GetReplacedMethods())
+                            {
+                                CompileMethod(replaced, extendedOrdinal, ref processedInitializers, synthesizedSubmissionFields, compilationState);
```
The code gen might be more stable (e.g. for edit-and-continue scenarios) if the ordinal were per method rather than for the whole enclosing type.

gafter replied:
Wait... how is this ordinal used?

cston replied:
Added a comment to `extendedOrdinal` that it is used to ensure replaced methods have unique ids.

gafter commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -1462,6 +1418,60 @@ private void ParseModifiers(SyntaxListBuilder tokens)
+            // the reset point.
+            ResetPoint resetPoint = GetResetPoint();
+
+            this.EatToken(); //move past "async"
```
"async" or "replace"

cston replied:
Changed to `contextual keyword`.

gafter commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -1451,70 +1455,31 @@ private void CheckMemberNameConflicts(DiagnosticBag diagnostics)
-
-                        SourceUserDefinedConversionSymbol previousConversion;
-                        if (conversionsAsConversions.TryGetValue(conversion, out previousConversion))
+                        if (!conversionsBySignature.Add(conversion))
```
Was the `conversionsAsConversions` stuff unnecessary?

cston replied:
I believe method signature collisions can be handled by `methodsBySignature` alone. As a result, for consistency, `conversionsAsConversions` was renamed to `conversionsBySignature`.

gafter commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -1451,70 +1455,31 @@ private void CheckMemberNameConflicts(DiagnosticBag diagnostics)
                     // second and third categories as follows:
 
                     var conversion = symbol as SourceUserDefinedConversionSymbol;
```
(I know you didn't write this line of code but) is there a test that can accomplish this semantically without depending on the specific implementation hierarchy?

cston replied:
Perhaps this can be accomplished by comparing ```symbol.Kind == SymbolKind.Method && (MethodSymbol)symbol).MethodKind == MethodKind.Conversion``` and then declaring `conversionsBySignature` as a dictionary of `MethodSymbol`. Will leave as is for now though.

gafter commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -1451,70 +1455,31 @@ private void CheckMemberNameConflicts(DiagnosticBag diagnostics)
-                            // we see a colliding method later.
-                            methodsBySignature[method] = method;
-                        }
+                foreach (var group in methodsBySignature.Values)
```
I don't understand why you separated the detection of the conflict (by producing multiple values into a `OneOrMany`) from the reporting. This appears to have the same behavior as before but at the cost of storing more transient data.

cston replied:
Good catch, thanks. This was left over from an earlier attempt to match up replaced methods. I've reverted the change.

gafter commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -2326,26 +2289,105 @@ internal Binder GetBinder(CSharpSyntaxNode syntaxNode)
+                        var method = member as SourceMethodSymbol;
+                        if ((object)method != null && method.IsPartial)
+                        {
+                            continue; // PROTOTYPE: Allow replacing partial methods.
```
All the `PROTOTYPE` comments should include `PROTOTYPE(generators)`, please.

cston replied:
Even though `PROTOTYPE` comments should be removed before merging into the future branch?

gafter replied:
Yes. It is likely we'll integrate early to facilitate testing multiple featured together by end users before every eye is crossed and tee dotted.

cston replied:
Changed all to ```PROTOTYPE(generators)```.

gafter commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -2411,15 +2453,19 @@ private static ImmutableArray<Symbol> FixPartialMember(ImmutableArray<Symbol> sy
+            return Remove(symbols, implementation);
+        }
+
+        private static ImmutableArray<Symbol> Remove(ImmutableArray<Symbol> symbols, Symbol symbol)
```
Given that replace/original may be used more than partial methods, doing this over and over again can be inefficient (i.e. quadratic). Does it make sense to `RemoveAll` a set of symbols rather than one at a time? (a PROTOTYPE comment or git issue would be a good idea to investigate this later).

cston replied:
Added a `PROTOTYPE` comment. Note this would only allow removing all replace symbols at once that have the same name.

gafter commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs:
```diff
@@ -669,6 +678,33 @@ internal sealed override OverriddenOrHiddenMembersResult OverriddenOrHiddenMembe
+            {
+                return (object)_replacedBy == null ?
+                    base.MetadataName :
+                    GeneratedNames.ReplacedMemberName(Name, uniqueId: 0);
```
I'm a bit confused about how the replaced and originals are linked together. Are they like a linked list, where each replace method links to and from the method it replaces, or does the original method have a list of all of the replacements, or both? Is it worth adding a comment summarizing the strategy somewhere?

cston replied:
The `Replaced` and `ReplacedBy` members reference each other. The `ReplacedBy` member is included in ```GetMembers()``` on the containing type while the `Replaced` member is only available from `ReplacedBy.Replaced`.

Add comments to `Symbol.Replaced` and `Symbol.ReplacedBy`.

cston replied:
The `Replaced` and `ReplacedBy` members reference each other. The `ReplacedBy` member is included in `GetMembers()` of the containing type while the `Replaced` member is only available from `ReplacedBy.Replaced`.

Added comments to `Symbol.Replaced` and `Symbol.ReplacedBy`.

gafter commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1609,21 +1612,14 @@ internal void EnsureAnonymousTypeTemplates(CancellationToken cancellationToken)
-            EmitStreamProvider pdbStreamProvider,
-            EmitStreamProvider xmlDocumentationStreamProvider,
-            EmitStreamProvider win32ResourcesStreamProvider,
+        internal CommonPEModuleBuilder PreCompile(
```
It would be nice to have some documentation on the relative role of PreCompile, Compile, and PostCompile. Just a sentence of two on each, please.

cston replied:
Added comments.

gafter commented:
LGTM aside from the few comments I made.
